From: Dmitry Bogatov <KAction@debian.org>
Date: Mon, 15 Jul 2019 23:51:43 +0000, 15 Sep 2019 23:45:24 +0000
Subject: Make runsvdir(8) rescan directory on SIGALARM + fix 1 second delay in forced-rescan implementation

This patch squash patches 0023 into patch 0019
Ref: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=933078
Ref: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=933078#40

Generally runsvdir(8) rescans service directory every 5 seconds. This patch
makes it possbile to force rescan by sending SIGALARM.

This feature is wanted by maintainer scripts of packages that want to perform
some action after service was started. Obliviously, hanging installation
process for 5 seconds is sub-optimal in such situation.
---
 src/Makefile   |  4 ++--
 src/runsvdir.c | 38 +++++++++++++++++++++++++++++++++++++--
 2 files changed, 39 insertions(+), 3 deletions(-)

--- a/src/Makefile
+++ b/src/Makefile
@@ -14,8 +14,8 @@
 runsv: load runsv.o unix.a byte.a time.a
 	./load runsv unix.a byte.a time.a
 
-runsvdir: load runsvdir.o unix.a byte.a time.a
-	./load runsvdir unix.a byte.a time.a
+runsvdir: load runsvdir.o unix.a fmt_ptime.o byte.a time.a byte.a
+	./load runsvdir unix.a fmt_ptime.o byte.a time.a byte.a
 
 runsvstat: load runsvstat.o unix.a byte.a time.a
 	./load runsvstat unix.a byte.a time.a
--- a/src/runsvdir.c
+++ b/src/runsvdir.c
@@ -15,10 +15,17 @@
 #include "iopause.h"
 #include "sig.h"
 #include "ndelay.h"
+#include "fmt_ptime.h"
 
 #define USAGE " [-P] dir"
 #define VERSION "$Id: 28cb04495cffccf27650de3cf2981942679781cd $"
 
+/*
+ * After rescan of {svdir} that was forced by sig_alarm (as opposed to regular
+ * rescan once every 5 seconds) is finished, timestamp is written to this path
+ * (relative to {svdir}).
+ */
+#define FORCED_RESCAN "./.forced-rescan"
 #define MAXSERVICES 1000
 
 char *progname;
@@ -40,6 +47,7 @@
 struct taia stamplog;
 int exitsoon =0;
 int pgrp =0;
+int siga =0;
 
 void usage () { strerr_die4x(1, "usage: ", progname, USAGE, "\n"); }
 void fatal(char *m1, char *m2) {
@@ -53,6 +61,28 @@
 } 
 void s_term() { exitsoon =1; }
 void s_hangup() { exitsoon =2; }
+void s_alarm() { siga = 1; }
+
+void report_forced_rescan() {
+  struct taia stamp;
+  char buf[26];
+  int fd;
+
+  taia_now(&stamp);
+  fmt_taia(buf, &stamp);
+  buf[25] =0;
+
+  fd =open_trunc(FORCED_RESCAN);
+  if (fd == -1) {
+    warn3x("failed to open", FORCED_RESCAN, buf);
+    return;
+  }
+
+  if (write(fd, buf, 25) != 25)
+    warn3x("failed to write", FORCED_RESCAN, buf);
+
+  close(fd);
+}
 
 void runsv(int no, char *name) {
   int pid;
@@ -70,6 +100,7 @@
     prog[2] =0;
     sig_uncatch(sig_hangup);
     sig_uncatch(sig_term);
+    sig_uncatch(sig_alarm);
     if (pgrp) setsid();
     pathexec_run(*prog, prog, (char* const*)environ);
     fatal("unable to start runsv ", name);
@@ -183,6 +214,7 @@
 
   sig_catch(sig_term, s_term);
   sig_catch(sig_hangup, s_hangup);
+  sig_catch(sig_alarm, s_alarm);
   svdir =*argv++;
   if (argv && *argv) {
     rplog =*argv;
@@ -225,7 +257,7 @@
       taia_add(&stampcheck, &now, &deadline);
       
       if (stat(svdir, &s) != -1) {
-        if (check || \
+        if (check || siga || \
             s.st_mtime != mtime || s.st_ino != ino || s.st_dev != dev) {
           /* svdir modified */
           if (chdir(svdir) != -1) {
@@ -233,9 +265,13 @@
             dev =s.st_dev;
             ino =s.st_ino;
             check =0;
-            if (now.sec.x <= (4611686018427387914ULL +(uint64)mtime))
+            if (!siga && now.sec.x <= (4611686018427387914ULL +(uint64)mtime))
               sleep(1);
             runsvdir();
+            if (siga) {
+              report_forced_rescan();
+              siga =0;
+            }
             while (fchdir(curdir) == -1) {
               warn("unable to change directory, pausing", 0);
               sleep(5);
